Use micromamba to bootstrap mamba

maresb opened the pull request:
If Micromamba is as mature as I think, then there is no longer any need to use Miniforge.

In case there are significant bugs in Micromamba, then maybe we find them here.

:eyes:

consideRatio commented:
@maresb I'm overall a bit confused about things, but I know that when I install `Mambaforge` release of the miniforge repo, i see "micromamba" logo show up. So, if it comes down to using mambaforge releases vs installing micromamba directly, is there a clear difference?

manics commented:
FYI we recently merged a micromamba PR https://github.com/jupyterhub/repo2docker/pull/1062 into repo2docker

maresb commented:
@consideRatio, Mambaforge is a 103 MB distribution containing conda/mamba. On the other hand, Micromamba is a 14MB binary which can be used to directly create an arbitrary environment directly from conda-forge.

maresb commented:
I think micromamba is one of the coolest programs I've ever seen. But if I get this PR to pass, there's still a legit discussion whether or not we should use it in favor of Miniforge. One could argue that maybe micromamba is not yet stable enough, but at least it is under very active development.

As for the current workflows, I'm not sure why wget/bzip is failing on "Build, test, and publish Docker Images":

```bash
  #12 [linux/arm64  6/13] RUN set -x &&     cd /usr/local &&     arch=$(uname -m) &&     if [ "${arch}" == "x86_64" ]; then         arch="64";     fi &&     wget -qO- "https://micromamba.snakepit.net/api/micromamba/linux-${arch}/latest" | tar -xvj bin/micromamba &&     micromamba create --prefix="/opt/conda"
  #12 0.100 + cd /usr/local
  #12 0.108 ++ uname -m
  #12 0.143 + arch=aarch64
  #12 0.147 + '[' aarch64 == x86_64 ']'
  #12 0.151 + tar -xvj bin/micromamba
  #12 0.152 + wget -qO- https://micromamba.snakepit.net/api/micromamba/linux-aarch64/latest
  #12 1.164 
  #12 1.164 bzip2: Compressed file ends unexpectedly;
  #12 1.164 	perhaps it is corrupted?  *Possible* reason follows.
  #12 1.166 bzip2: Inappropriate ioctl for device
  #12 1.167 	Input file = (stdin), output file = (stdout)
```

That Tensorflow is failing to solve in the other workflow is a bit less surprising to me. :laughing:

consideRatio commented:
Ah, so by using miniforge we would get a somewhat delayed version of python, while using micromamba we get the absolute latest unless we explicitly pin it down. With that said, using micromamba we will need a policy regarding upgrading the Python version. In my mind, we should not switch to Python 3.10 because its too new, but what policy to follow? Using what ships with miniforge releases seems reasonable to me.

maresb commented:
Yes, exactly. It provides more freedom and the corresponding responsibility. Another possible policy could be to use the latest Python where everything works. Tensorflow is an extremely complicated package, so if that's running, the corresponding Python version must be very stable.

I'm thinking that I should combine the download of Micromamba with the mamba install so that I can delete Micromamba afterwards and save the 14MB. Micromamba is redundant when Mamba is installed.

maresb commented:
I'm baffled by the bzip2 failure on aarch64. The file seems to extract locally for me. Anyone have an idea?

mathbunnyru commented:
Unfortunately, I have no idea what's wrong with bzip2.
Have you tried inside aarch64 docker?

maresb commented:
@mathbunnyru ya, following [this](https://www.stereolabs.com/docs/docker/building-arm-container-on-x86/) I did
```bash
sudo apt-get install qemu binfmt-support qemu-user-static
docker run --rm --privileged multiarch/qemu-user-static --reset -p yes
docker run --rm -it aarch64/ubuntu
uname -m
apt-get update && apt-get install -y wget bzip2
wget -qO- "https://micromamba.snakepit.net/api/micromamba/linux-aarch64/latest" | tar -xvj bin/micromamba
micromamba create --prefix=/opt/conda
```
and it runs fine! :thinking:

mathbunnyru commented:
Let's try it once again then :)

maresb commented:
It's already failed multiple times

mathbunnyru commented:
Oh, I see :(

maresb commented:
I'll try turning on verbose on wget...

maresb commented:
That was instructive. On GitHub,

```
   #11 1.250 ---request begin---
  #11 1.250 GET /download/conda-forge/micromamba/0.20.0/linux-aarch64/micromamba-0.20.0-0.tar.bz2 HTTP/1.1
  #11 1.250 User-Agent: Wget/1.20.3 (linux-gnu)
  #11 1.250 Accept: */*
  #11 1.250 Accept-Encoding: identity
  #11 1.250 Host: api.anaconda.org
  #11 1.250 Connection: Keep-Alive
  #11 1.250 
  #11 1.250 ---request end---
  #11 1.659 
  #11 1.659 bzip2: Compressed file ends unexpectedly;
  #11 1.659 	perhaps it is corrupted?  *Possible* reason follows.
  #11 1.660 bzip2: Inappropriate ioctl for device
  #11 1.660 	Input file = (stdin), output file = (stdout)
```

Locally, after that last `--request end---` line, I see

```
---response begin---
HTTP/1.1 302 FOUND
Date: Wed, 02 Feb 2022 20:33:42 GMT
Content-Type: text/html; charset=utf-8
```

So then the question is what to do about it.

I notice that the User-Agent headers are slightly different. Locally I have `Wget/1.21` but on GitHub there is `Wget/1.20.3 (linux-gnu)`. Maybe that indicates some strange discrepancy?

I suppose I could try with `curl`, but that's another package to install. :man_shrugging:

maresb commented:
Fun fun...

Wget is causing a [segfault](https://github.com/jupyter/docker-stacks/runs/5043958357?check_suite_focus=true#step:8:1473).

Also curl is causing a [segfault](https://github.com/jupyter/docker-stacks/runs/5043703775?check_suite_focus=true#step:8:1431).

This is quite impressive. Here's the version info:

```
  #13 0.219 + uname -a
  #13 0.255 Linux buildkitsandbox 5.11.0-1028-azure #31~20.04.2-Ubuntu SMP Tue Jan 18 08:46:15 UTC 2022 aarch64 aarch64 aarch64 GNU/Linux
  #13 0.260 + echo aarch64
  #13 0.262 aarch64
  #13 0.262 + wget --version
  #13 0.320 GNU Wget 1.20.3 built on linux-gnu.
  #13 0.320 
  #13 0.320 -cares +digest -gpgme +https +ipv6 +iri +large-file -metalink +nls 
  #13 0.320 +ntlm +opie +psl +ssl/openssl 
  #13 0.320 
  #13 0.320 Wgetrc: 
  #13 0.320     /etc/wgetrc (system)
  #13 0.320 Locale: 
  #13 0.320     /usr/share/locale 
  #13 0.320 Compile: 
  #13 0.320     gcc -DHAVE_CONFIG_H -DSYSTEM_WGETRC="/etc/wgetrc" 
  #13 0.320     -DLOCALEDIR="/usr/share/locale" -I. -I../../src -I../lib 
  #13 0.320     -I../../lib -Wdate-time -D_FORTIFY_SOURCE=2 -DHAVE_LIBSSL -DNDEBUG 
  #13 0.320     -g -O2 -fdebug-prefix-map=/build/wget-1R1vjg/wget-1.20.3=. 
  #13 0.320     -fstack-protector-strong -Wformat -Werror=format-security 
  #13 0.320     -DNO_SSLv2 -D_FILE_OFFSET_BITS=64 -g -Wall 
  #13 0.320 Link: 
  #13 0.320     gcc -DHAVE_LIBSSL -DNDEBUG -g -O2 
  #13 0.320     -fdebug-prefix-map=/build/wget-1R1vjg/wget-1.20.3=. 
  #13 0.320     -fstack-protector-strong -Wformat -Werror=format-security 
  #13 0.320     -DNO_SSLv2 -D_FILE_OFFSET_BITS=64 -g -Wall -Wl,-Bsymbolic-functions 
  #13 0.320     -Wl,-z,relro -Wl,-z,now -lpcre2-8 -luuid -lidn2 -lssl -lcrypto -lz 
  #13 0.320     -lpsl ftp-opie.o openssl.o http-ntlm.o ../lib/libgnu.a 
  #13 0.320 
  #13 0.320 Copyright (C) 2015 Free Software Foundation, Inc.
  #13 0.320 License GPLv3+: GNU GPL version 3 or later
  #13 0.320 <http://www.gnu.org/licenses/gpl.html>.
  #13 0.320 This is free software: you are free to change and redistribute it.
  #13 0.320 There is NO WARRANTY, to the extent permitted by law.
  #13 0.320 
  #13 0.320 Originally written by Hrvoje Niksic <hniksic@xemacs.org>.
  #13 0.320 Please send bug reports and questions to <bug-wget@gnu.org>.
  #13 0.324 + curl --version
  #13 0.422 curl 7.68.0 (aarch64-unknown-linux-gnu) libcurl/7.68.0 OpenSSL/1.1.1f zlib/1.2.11 brotli/1.0.7 libidn2/2.2.0 libpsl/0.21.0 (+libidn2/2.2.0) libssh/0.9.3/openssl/zlib nghttp2/1.40.0 librtmp/2.3
  #13 0.422 Release-Date: 2020-01-08
  #13 0.422 Protocols: dict file ftp ftps gopher http https imap imaps ldap ldaps pop3 pop3s rtmp rtsp scp sftp smb smbs smtp smtps telnet tftp 
  #13 0.422 Features: AsynchDNS brotli GSS-API HTTP2 HTTPS-proxy IDN IPv6 Kerberos Largefile libz NTLM NTLM_WB PSL SPNEGO SSL TLS-SRP UnixSockets
```

maresb commented:
I finally managed to reproduce this locally after following the instructions in my previous comment and running:
```
docker run --rm -it --platform "linux/arm64" ubuntu:focal@sha256:669e010b58baf5beb2836b253c1fd5768333f0d1dbcb834f7c07a4dc93f474be
apt-get update && apt-get install wget curl
wget --debug "https://api.anaconda.org/download/conda-forge/micromamba/0.20.0/linux-aarch64/micromamba-0.20.0-0.tar.bz2"
curl --verbose -o file "https://api.anaconda.org/download/conda-forge/micromamba/0.20.0/linux-aarch64/micromamba-0.20.0-0.tar.bz2"
```
Then I get
```
root@3decd724b8f4:/# strace wget --debug "https://api.anaconda.org/download/conda-forge/micromamba/0.20.0/linux-aarch64/micromamba-0.20.0-0.tar.bz2"
/usr/bin/strace: test_ptrace_get_syscall_info: PTRACE_TRACEME: Function not implemented
/usr/bin/strace: ptrace(PTRACE_TRACEME, ...): Function not implemented
/usr/bin/strace: PTRACE_SETOPTIONS: Function not implemented
/usr/bin/strace: detach: waitpid(7782): No child processes
/usr/bin/strace: Process 7782 detached
root@3decd724b8f4:/# wget --debug "https://api.anaconda.org/download/conda-forge/micromamba/0.20.0/linux-aarch64/micromamba-0.20.0-0.tar.bz2"
DEBUG output created by Wget 1.20.3 on linux-gnu.

Reading HSTS entries from /root/.wget-hsts
URI encoding = 'ANSI_X3.4-1968'
iconv UTF-8 -> ANSI_X3.4-1968
iconv outlen=210 inlen=105
converted 'https://api.anaconda.org/download/conda-forge/micromamba/0.20.0/linux-aarch64/micromamba-0.20.0-0.tar.bz2' (ANSI_X3.4-1968) -> 'https://api.anaconda.org/download/conda-forge/micromamba/0.20.0/linux-aarch64/micromamba-0.20.0-0.tar.bz2' (UTF-8)
Converted file name 'micromamba-0.20.0-0.tar.bz2' (UTF-8) -> 'micromamba-0.20.0-0.tar.bz2' (ANSI_X3.4-1968)
--2022-02-02 23:58:26--  https://api.anaconda.org/download/conda-forge/micromamba/0.20.0/linux-aarch64/micromamba-0.20.0-0.tar.bz2
Resolving api.anaconda.org (api.anaconda.org)... 104.17.93.24, 104.17.92.24, 2606:4700::6811:5d18, ...
Caching api.anaconda.org => 104.17.93.24 104.17.92.24 2606:4700::6811:5d18 2606:4700::6811:5c18
Connecting to api.anaconda.org (api.anaconda.org)|104.17.93.24|:443... connected.
Created socket 3.
Releasing 0x00000055000bd710 (new refcount 1).
Initiating SSL handshake.
Handshake successful; connected socket 3 to SSL handle 0x00000055000bd890
certificate:
  subject: CN=anaconda.org,O=Cloudflare\\, Inc.,L=San Francisco,ST=California,C=US
  issuer:  CN=Cloudflare Inc ECC CA-3,O=Cloudflare\\, Inc.,C=US
X509 certificate successfully verified and matches host api.anaconda.org

---request begin---
GET /download/conda-forge/micromamba/0.20.0/linux-aarch64/micromamba-0.20.0-0.tar.bz2 HTTP/1.1
User-Agent: Wget/1.20.3 (linux-gnu)
Accept: */*
Accept-Encoding: identity
Host: api.anaconda.org
Connection: Keep-Alive

---request end---
Segmentation fault (core dumped)
```
I don't know how to debug the segfault though.

mathbunnyru commented:
I don't know what to suggest here.
I only know kind of dirty fix - use multi-stage build and simply use a different image to just download `micromamba` executable and then copy that executable from this image.
I think you might get away with `alpine` for example.

This way you can also install curl/wget/bzip2 and whatever you want in your alpine image and it won't affect the final image.

maresb commented:
In my aarch64 Docker container, I compiled wget 1.21.2 from source and I can download successfully with the compiled binary but the original system binary still segfaults.

```
docker run --rm -it --platform "linux/arm64" ubuntu:focal@sha256:669e010b58baf5beb2836b253c1fd5768333f0d1dbcb834f7c07a4dc93f474be
DEBIAN_FRONTEND=noninteractive TZ=Etc/UTC apt-get update && apt-get install -y wget build-essential pkg-config gnutls-dev
wget https://ftp.gnu.org/gnu/wget/wget-latest.tar.gz
tar -xf wget-latest.tar.gz
cd wget-1.21.2/
./configure
make
# Succeeds:
src/wget --debug "https://api.anaconda.org/download/conda-forge/micromamba/0.20.0/linux-aarch64/micromamba-0.20.0-0.tar.bz2"
# Fails:
wget --debug "https://api.anaconda.org/download/conda-forge/micromamba/0.20.0/linux-aarch64/micromamba-0.20.0-0.tar.bz2"
```

I made a quick attempt at your suggestion, and now Docker seems very confused about platforms. It's now defaulting to `linux/arm64` for `ubuntu:focal`. I think I need to clean up my images. This is my first time doing multiplatform so I'm a bit confused. It's saying `The requested image's platform (linux/arm64/v8) does not match the detected host platform (linux/amd64) and no specific platform was requested` even though I'm building with `--platform=linux/amd64`. Maybe I have to remove the image and redownload?

mathbunnyru commented:
Unfortunately, I don't have enough experience with multi-platform docker images to answer your question, sorry :(

mathbunnyru commented:
I have checked the docs:
https://docs.docker.com/buildx/working-with-buildx/#build-multi-platform-images

There is an interesting example:
```
# syntax=docker/dockerfile:1
FROM --platform=$BUILDPLATFORM golang:alpine AS build
ARG TARGETPLATFORM
ARG BUILDPLATFORM
RUN echo "I am running on $BUILDPLATFORM, building for $TARGETPLATFORM" > /log
FROM alpine
COPY --from=build /log /log
```

If you only use the stage to download something, you can actually run it using the build platform, which means it won't use QEMU, will run faster, and will only be pulled once.

mathbunnyru commented on base-notebook/Dockerfile:
```diff
@@ -5,6 +5,23 @@
+        arch="64"; \
+    fi && \
+    echo $arch && \
+    apk add --no-cache curl==7.80.0-r0 && \
```
Could we not specify curl version here?

maresb replied:
Thanks for the review! I only did it to prevent hadolint from complaining. I also think it's weird; I'll add an ignore instead.

mathbunnyru replied:
Thanks! I thought we explicitly asked hadolint to ignore this rule be default in our project.

mathbunnyru commented on base-notebook/Dockerfile:
```diff
@@ -5,6 +5,23 @@
+    fi && \
+    echo $arch && \
+    apk add --no-cache curl==7.80.0-r0 && \
+    curl -L "https://micromamba.snakepit.net/api/micromamba/linux-${arch}/latest" > micromamba.tar.bz2
```
```suggestion
    curl -L "https://micromamba.snakepit.net/api/micromamba/linux-${arch}/latest" -o /tmp/micromamba.tar.bz2
```

And I think WORKDIR can be avoided.

maresb replied:
Also on hadolint's suggestion :joy:

mathbunnyru replied:
Ok, let's keep it then

maresb replied:
I'm making the commit now (to a separate branch to not disturb the CI). Using `WORKDIR` was hadolint, I'm going to use `-o`.

Ah! I didn't understand the causality here. Good point: no workdir needed, I think.

mathbunnyru commented on base-notebook/Dockerfile:
```diff
@@ -40,7 +57,10 @@ RUN apt-get update --yes && \
     tini \
-    wget && \
+    wget \
+    curl \
```
```suggestion
```

maresb replied:
Ya, this is a bit of a rough draft. I'd like to squash, assuming I get it running. If you approve, of course. I don't want to force push too much. :wink:

mathbunnyru replied:
It's ok if you don't squash. I don't like push forces, especially when someone already reviewed the content.
Just keep it the way it is, it's ok for us to have "Fix this" and "Fix that" comments.

mathbunnyru commented on base-notebook/Dockerfile:
```diff
@@ -69,71 +89,64 @@ RUN sed -i 's/^#force_color_prompt=yes/force_color_prompt=yes/' /etc/skel/.bashr
-    /bin/bash "${miniforge_installer}" -f -b -p "${CONDA_DIR}" && \
-    rm "${miniforge_installer}" && \
+    tar -xvjf micromamba.tar.bz2 bin/micromamba && \
+    # bin/micromamba create --prefix="${CONDA_DIR}" && \
```
Should we delete this?

mathbunnyru reviewed (changes requested):
Nice! I hope this works as well :)

mathbunnyru commented:
That's actually what I intended you to test, so there was a problem in your local cache, right?

maresb commented:
> That's actually what I wanted you to try, so there was a problem in your local cache, right?

Ya, I was testing using `docker build` instead of `docker buildx build`.

The way I was able to get it working locally was:
```bash
docker buildx build --platform=linux/arm64 .
```

(with of course the commands from my earlier comments)

maresb commented on base-notebook/Dockerfile:
```diff
@@ -67,73 +69,67 @@ RUN sed -i 's/^#force_color_prompt=yes/force_color_prompt=yes/' /etc/skel/.bashr
    # Add call to conda init script see https://stackoverflow.com/a/58081608/4413446
    echo 'eval "$(command conda shell.bash hook 2> /dev/null)"' >> /etc/skel/.bashrc
 
+# Download and install Micromamba, and initialize Conda prefix.
```
Sorry I didn't pay more attention to this review since I was focused on getting it running. Is this still relevant?

mathbunnyru commented:
I'm completely lost here. Why new PR? Should I review something or not?

maresb commented:
I didn't want to interrupt this CI run, but I still wanted to make changes. I plan to close the other PR after merging the changes here.

It's up to you whether you feel like reviewing the other one. (It's identical except for the latest commits.) Maybe it's easier to just wait until I merge them here.

Do you think this run is frozen? Or is it just slow?

mathbunnyru commented:
Ok, I will wait, no problem.
It's already taking almost 2 hours longer than the usual build, so it's probably stuck :(

maresb commented:
Hmm, so here with a99cd6d the [action](https://github.com/jupyter/docker-stacks/runs/5056938642?check_suite_focus=true) stalled right away on base-notebook. On my other branch with d77c32f the [action](https://github.com/jupyter/docker-stacks/runs/5059391145?check_suite_focus=true) stalled on scipy-notebook. Is this #1539 again?

By the way, I wonder if it's possible to simplify your GitHub Action workflow. Maybe not because we have so many images to build here, but it might be worthwhile to compare with [micromamba-docker](https://github.com/mamba-org/micromamba-docker/blob/main/.github/workflows/push_latest.yml).

maresb commented:
Was #1539 a sporadic issue? Maybe I thought that #1593 solved it, but instead I just got lucky?

mathbunnyru commented:
> Was #1539 a sporadic issue? Maybe I thought that #1593 solved it, but instead I just got lucky?

Yes, you got lucky. I reverted the commit and reopened the issue. So, please, merge the latest master to your branch.

mathbunnyru commented on base-notebook/Dockerfile:
```diff
@@ -80,72 +100,54 @@ RUN echo "auth requisite pam_deny.so" >> /etc/pam.d/su && \
-    'jupyterhub' \
-    'jupyterlab' && \
+COPY --chown="${NB_UID}:${NB_GID}" --from=download_micromamba_stage /tmp/micromamba.tar.bz2 /tmp
+COPY --chown="${NB_UID}:${NB_GID}"  initial-condarc "${CONDA_DIR}/.condarc"
```
```suggestion
COPY --chown="${NB_UID}:${NB_GID}" initial-condarc "${CONDA_DIR}/.condarc"
```

maresb replied:
Thanks!

mathbunnyru commented on base-notebook/Dockerfile:
```diff
@@ -80,72 +100,54 @@ RUN echo "auth requisite pam_deny.so" >> /etc/pam.d/su && \
+        MAMBA_SPECIFIER='mamba<0.18'; \
+    fi && \
+    # Install the packages
+    /tmp/micromamba install \
```
In this case I would probably change WORKDIR to /tmp, because having many `/tmp/` prefixes is not beatiful.
In the end it's a matter of personal choice, I beleive.

maresb commented:
By the way, if we go this route, I think it would be more convenient to use micromamba-docker as the base image.

Unfortunately it is not yet suitable because they use [Debian instead of Ubuntu](https://github.com/mamba-org/micromamba-docker/issues/59).

mathbunnyru commented:
I have checked the [dockerfile](https://github.com/mamba-org/micromamba-docker/blob/main/Dockerfile) and I've noticed several features that we should incorporate (I will probably do it later):

`FROM --platform=$BUILDPLATFORM $BASE_IMAGE AS stage1` - this one is exactly as I suggested above
`RUN test "$TARGETARCH" = 'amd64' && export ARCH='64';` - looks better than our `if statements`
`SHELL` command - I know nothing about this feature, I need to check what that is

maresb commented:
The `SHELL` command is what is used to interpret the `RUN` statements in the Dockerfile. It defaults to `sh` but for our purposes it's better to switch to `bash` because Conda environments don't support `sh`.

maresb commented:
It looks like this is hanging despite reverting #1593.

mathbunnyru commented:
Canceled the workflow.
Now it seems that `micromamba` is stuck.
Maybe now we have to use old micromamba as well?

maresb commented:
I really don't understand what could be causing the hang. Since I'm almost out of ideas, I'm feeling really close to calling on the mamba devs for help here. But I want to make sure we exhaust all possible ideas that we have first.

maresb commented:
This one is really messing with my head!!! I'm actually frustrated now that it's passing! :angry:

How can it possibly be that having run micromamba once somehow dooms mamba to fail?

I'm going to restart CI to get another data point...

maresb commented:
It passed yet again, but is this just luck???

Here's the current status:

micromamba=latest, mamba=latest, fail.
micromamba=latest, mamba=0.17.0, fail.
micromamba=0.17.0, mamba=0.17.0, success.
micromamba=0.17.0, mamba=0.17.0, success again.
micromamba=latest, mamba=0.17.0, pending.

(Please ignore my commit "Downgrade micromamba again", I miswrote my commit message, and then immediately force-pushed a corrected commit.)

mathbunnyru commented:
I think `micromamba=0.17.0, mamba=0.17.0` is the only one that's going to work.
There seems to be a huge change in the whole mamba project after 0.17.
That's why it kind of makes sense that if the new mamba doesn't work, then micromamba is also stuck - they share a lot of the codebase I suppose.

maresb commented:
What really bothers me is that micromamba is a statically linked executable which we delete. There is zero shared code. It really should not matter what version of micromamba we had used once it is deleted.

mathbunnyru commented:
@maresb also, if you would like to proceed with this, please, make sure, that docs stay up-to-date (grep `Miniforge`, `Mambaforge` and similar keywords).

maresb commented:
@mathbunnyru will do. But I'm not yet ready to proceed until I understand why it sometimes freezes.

mathbunnyru commented:
@maresb it seems that now latest versions work quite well? :)

mathbunnyru commented:
`0.21.1` was just released - I will check it works too.

mathbunnyru commented:
> @mathbunnyru, seems that you spoke too soon. 😢
> 
> I'm still tracking all the runs here: [#1594 (comment)](https://github.com/jupyter/docker-stacks/pull/1594#issuecomment-1030333562)
> 
> I feel like we have gathered lots of data and are out of ideas, so I think I'm justified in calling for reinforcements. Let's ask this weekend the author of Micromamba if he has any ideas.

Thanks @maresb 👍

maresb commented:
Micromamba was set to "quiet" and i think it would be good to know at what stage it is freezing, so i enabled verbose. Now I just need to reproduce the freeze.

maresb commented:
@mathbunnyru I'm at a loss on what to do here. I want to have the verbose logs in hand before asking for help, but now I can't reproduce the problem.

I get the impression that the CI runs are reproducible but still susceptible to some sort of random perturbations. Would it be okay if I rebase onto the latest master in order to give it a little kick?

mathbunnyru commented:
@maresb ok! 👍

mathbunnyru commented:
@maresb I've just checked and the workaround with downloading in another docker image is not needed anymore.

https://github.com/jupyter/docker-stacks/pull/1594#issuecomment-1028472495

This works fine for me (but of course I'm using latest `ubuntu:focal` version)

maresb commented:
@mathbunnyru, running `docker pull ubuntu:focal` seemed to resolve the segfaults on my side. Seems like it was likely a bug. Let's see how it works now...

mathbunnyru commented:
The build step worked well, so at least this hack is not needed anymore.

mathbunnyru commented on base-notebook/Dockerfile:
```diff
@@ -80,72 +82,64 @@ RUN echo "auth requisite pam_deny.so" >> /etc/pam.d/su && \
+        # Should be simpler, see <https://github.com/mamba-org/mamba/issues/1437>
+        arch="64"; \
+    fi && \
+    echo $arch && \
+    # curl -L "https://micromamba.snakepit.net/api/micromamba/linux-${arch}/${MICROMAMBA_VERSION}" \
+    #     -o /tmp/micromamba.tar.bz2 && \
```
```suggestion
```

consideRatio commented on base-notebook/Dockerfile:
```diff
@@ -80,72 +82,61 @@ RUN echo "auth requisite pam_deny.so" >> /etc/pam.d/su && \
-    'notebook' \
-    'jupyterhub' \
-    'jupyterlab' && \
+COPY --chown="${NB_UID}:${NB_GID}" initial-condarc "${CONDA_DIR}/.condarc"
```
I don't remember clearly, but I recall that `--chown` wasn't supported or behaved a bit differently - watch out for possible issues about that. I'm unable to recollect something in more detail than this though =/

maresb replied:
Hmm, at this point in the Dockerfile we are running as `NB_USER`, so without the chown, `fix-permissions "${CONDA_DIR}"` will fail on my `.condarc`.

I think `--chown` is ~2 years old, so back then it would fail in many cases... but I hope that it's well-supported now?

I think the alternative would be something like
```Dockerfile
USER root
RUN chown blah
USER $NB_UID
```

A bit ugly, but if you're that concerned we could do it.

consideRatio replied:
I'm fine with what you want to try, but I want to warn about the vague memory I have that there isn't parity between COPY --chown and `RUN chown ...`  steps.

maresb commented:
Any idea why the tests are failing? Seems odd. Is it Python 3.10?

mathbunnyru commented:
> Any idea why the tests are failing? Seems odd. Is it Python 3.10?

I tried to explain it here: https://github.com/jupyter/docker-stacks/pull/1675

mathbunnyru reviewed (commented):
This looks great to me.
What I want to ask for is to do the testing, and check that it's not stuck in `arm` as it did in the past.
After that, we can hopefully remove `MAMBA_SPECIFIER` and merge this

maresb commented:
@mathbunnyru, what exactly do you have in mind for:
> What I want to ask for is to do the testing, and check that it's not stuck in arm as it did in the past.

Like retrying over and over as before to make sure it doesn't get stuck? Or are you thinking of a unit test?

mathbunnyru commented:
@maresb I'm afraid retrying over and over is the only solution here :(
Because it is stuck quite rarely, it might happen on some machine and not the other one or something like that.

I've got an idea to make it easy for you.
Just allow github actions in your fork (if they are not yet allowrd) and create and push 10-20 branches (all pointing to the last commit here).
This way you will launch this action in parallel and don't have to wait till something finishes.
A bit messy, but still better than having to wait and then close/open the PR, I think.

mathbunnyru commented:
@maresb please, merge master here, the tests should be fine now.

maresb commented:
@mathbunnyru I followed your suggestion (great idea!), and I'm currently running 10 tests in parallel in maresb/docker-stacks.

maresb commented:
3/10 of the runs in my repos have stalled. I'm reenabling micromamba's verbose mode in the hopes of finally seeing exactly where it stalls.

maresb commented:
I managed to trigger a failure with the logs enabled on 2/10 of the runs from my repo, but the logs got truncated after 2MB, so I didn't see any relevant output. I just reduced the verbosity in the hope of capturing something.

maresb commented:
For unsuccessful runs, the logs stop at what I believe is the last extracted package:
```
  #29 182.4 debug    libmamba Decompressing '/opt/conda/pkgs/bleach-4.1.0-pyhd8ed1ab_0.tar.bz2'
  #29 182.4 debug    libmamba Running subprocess extraction '/tmp/micromamba package extract /opt/conda/pkgs/bleach-4.1.0-pyhd8ed1ab_0.tar.bz2 /opt/conda/pkgs/bleach-4.1.0-pyhd8ed1ab_0'
  #29 183.7 debug    libmamba Extracted to '/opt/conda/pkgs/bleach-4.1.0-pyhd8ed1ab_0'
  #29 183.7 debug    libmamba Decompressing '/opt/conda/pkgs/certipy-0.1.3-py_0.tar.bz2'
  #29 183.7 debug    libmamba Running subprocess extraction '/tmp/micromamba package extract /opt/conda/pkgs/certipy-0.1.3-py_0.tar.bz2 /opt/conda/pkgs/certipy-0.1.3-py_0'
  #29 185.0 debug    libmamba Extracted to '/opt/conda/pkgs/certipy-0.1.3-py_0'
  #29 185.0 debug    libmamba Decompressing '/opt/conda/pkgs/libuv-1.43.0-hf897c2e_0.tar.bz2'
  #29 185.0 debug    libmamba Running subprocess extraction '/tmp/micromamba package extract /opt/conda/pkgs/libuv-1.43.0-hf897c2e_0.tar.bz2 /opt/conda/pkgs/libuv-1.43.0-hf897c2e_0'
  #29 186.7 debug    libmamba Extracted to '/opt/conda/pkgs/libuv-1.43.0-hf897c2e_0'
  #29 186.7 debug    libmamba Decompressing '/opt/conda/pkgs/mistune-0.8.4-py39h14843e3_1005.tar.bz2'
  #29 186.7 debug    libmamba Running subprocess extraction '/tmp/micromamba package extract /opt/conda/pkgs/mistune-0.8.4-py39h14843e3_1005.tar.bz2 /opt/conda/pkgs/mistune-0.8.4-py39h14843e3_1005'
  #29 188.0 debug    libmamba Extracted to '/opt/conda/pkgs/mistune-0.8.4-py39h14843e3_1005'
```

For successful runs, the logs continue with the linking stage, for instance:

```
#29 178.0 Linking pandoc-2.17.1.1-h8af1aa0_0
#29 178.0 debug    libmamba Verify cache '/opt/conda/pkgs' for package extracted directory 'pandoc-2.17.1.1-h8af1aa0_0'
#29 178.0 debug    libmamba 'pandoc-2.17.1.1-h8af1aa0_0' extracted directory cache is valid
#29 178.0 debug    libmamba Linking package 'pandoc-2.17.1.1-h8af1aa0_0' from '/opt/conda/pkgs/pandoc-2.17.1.1-h8af1aa0_0'
#29 178.0 debug    libmamba 1 files linked
#29 178.0 debug    libmamba post-link script for 'pandoc' does not exist ('/opt/conda/bin/.pandoc-post-link.sh')
#29 178.0 debug    libmamba Finalizing linking
#29 178.0 Linking libstdcxx-ng-11.2.0-h0d0a5bb_14
#29 178.0 debug    libmamba Verify cache '/opt/conda/pkgs' for package extracted directory 'libstdcxx-ng-11.2.0-h0d0a5bb_14'
#29 178.0 debug    libmamba 'libstdcxx-ng-11.2.0-h0d0a5bb_14' extracted directory cache is valid
#29 178.0 debug    libmamba Linking package 'libstdcxx-ng-11.2.0-h0d0a5bb_14' from '/opt/conda/pkgs/libstdcxx-ng-11.2.0-h0d0a5bb_14'
#29 178.0 debug    libmamba 4 files linked
#29 178.0 debug    libmamba post-link script for 'libstdcxx-ng' does not exist ('/opt/conda/bin/.libstdcxx-ng-post-link.sh')
#29 178.0 debug    libmamba Finalizing linking
#29 178.0 debug    libmamba Verify cache '/opt/conda/pkgs' for package extracted directory 'ld_impl_linux-aarch64-2.36.1-h02ad14f_2'
#29 178.0 Linking ld_impl_linux-aarch64-2.36.1-h02ad14f_2
#29 178.0 debug    libmamba 'ld_impl_linux-aarch64-2.36.1-h02ad14f_2' extracted directory cache is valid
#29 178.0 debug    libmamba Linking package 'ld_impl_linux-aarch64-2.36.1-h02ad14f_2' from '/opt/conda/pkgs/ld_impl_linux-aarch64-2.36.1-h02ad14f_2'
#29 178.0 debug    libmamba 4 files linked
#29 178.0 debug    libmamba post-link script for 'ld_impl_linux-aarch64' does not exist ('/opt/conda/bin/.ld_impl_linux-aarch64-post-link.sh')
#29 178.0 debug    libmamba Finalizing linking
```

Now that I _finally_ have something concrete, I feel okay about calling in @wolfv.

@wolfv, we have been trying in this PR to switch from Miniforge to Micromamba. We have been plagued by sporadic hangs in our multi-arch Docker build process on GH Actions at a rate of ~20% of all runs when we `RUN` the command

```bash
./micromamba install --root-prefix=/opt/conda --prefix=/opt/conda -vv --yes python=3.9 mamba notebook jupyterhub jupyterlab
```

Execution hangs without any further output until the 6h limit is reached. Do you have any ideas about what could be causing this?

mathbunnyru commented:
Thank you for your patience @maresb.

Also, this problem is not specific to `micromamba`.
It also happens with mamba from version `0.18`.
That's why now we explicitly pin the `mamba` version for aarch64.
https://github.com/jupyter/docker-stacks/blob/master/base-notebook/Dockerfile#L131

maresb commented:
Based on https://github.com/mamba-org/mamba/issues/1611 I'm confident that we have a reliable workaround for the issue of Mamba hanging on some builds.

To be absolutely sure, I have some workflow runs finishing up [here](https://github.com/maresb/docker-stacks/actions/workflows/docker.yml).

Then there is the issue of advantages and disadvantages.

* Network bandwidth: trusting the displayed download size, this is 146 MB for Mambaforge and 188 MB for Micromamba (=89 for archive + 31 for update + 68 for Jupyter).
* Flexibility: strong advantage to Micromamba since we can pick our exact Python version
* Image size: almost identical, but Micromamba is about 6MB smaller, due to Python 3.9.12 being that much smaller (yes, smaller) than 3.9.10
* Complexity: to my eyes, both Mambaforge and Micromamba have similar complexity in the Dockerfile
* Speed: omitting the `apt-get` step, I measure 2m04s for Mambaforge vs 1m25s for Micromamba
* Reliability: I think Mambaforge is the clear winner here, but I'm confident that the current issue has been resolved
* Future-proof: I'm personally convinced that Micromamba is the tool of the future :man_astronaut:

mathbunnyru commented on base-notebook/Dockerfile:
```diff
@@ -81,72 +83,65 @@ RUN echo "auth requisite pam_deny.so" >> /etc/pam.d/su && \
+
+# Pin versions here, or set them to "default"
+ARG PYTHON_VERSION=3.9
+ARG MAMBA_VERSION=default
+ARG MICROMAMBA_VERSION=default
```
Is this flexibility needed?

maresb replied:
Awesome. The builds no longer hang.

Sure, we can hardcode the default versions. Would you like me to do it, or are you on it?

mathbunnyru replied:
I'll do it, no worries :)

mathbunnyru replied:
Done

mathbunnyru reviewed (approved):
Great news, @maresb!
I absolutely like this PR.

Waiting for your confirmation, that the builds are not stuck anymore, and then I think we can merge.

maresb commented:
Are we ready here? My tests show the builds no longer hanging. 🚀

mathbunnyru commented:
Yes, merging :)
